src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/NodesDnApiAction.java

```diff
@@ -0,0 +1,150 @@
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.rest.RestChannel;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestRequest.Method;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * This class implements CRUD operations to manage dynamic NodesDn. The primary usecase is targeted at cross-cluster where
+ * in node restart can be avoided by populating the coordinating cluster's nodes_dn values.
+ *
+ * The APIs are only accessible to SuperAdmin since the configuration controls the core application layer trust validation.
+ * By default the APIs are disabled and can be enabled by a YML setting - {@link ConfigConstants#OPENDISTRO_SECURITY_NODES_DN_DYNAMIC_CONFIG_ENABLED}
+ *
+ * The backing data is stored in {@link ConfigConstants#OPENDISTRO_SECURITY_CONFIG_INDEX_NAME} which is populated during bootstrap.
+ * For existing clusters, {@link com.amazon.opendistroforelasticsearch.security.tools.OpenDistroSecurityAdmin} tool can
+ * be used to populate the index.
+ *
+ * See {@link com.amazon.opendistroforelasticsearch.security.dlic.rest.api.NodesDnApiTest} for usage examples.
+ */
+public class NodesDnApiAction extends PatchableResourceApiAction {
+    public static final String STATIC_ES_YML_NODES_DN = "STATIC_ES_YML_NODES_DN";
+    private final List<String> staticNodesDnFromEsYml;
+
+    @Inject
+    public NodesDnApiAction(final Settings settings, final Path configPath, final RestController controller, final Client client,
+        final AdminDNs adminDNs, final ConfigurationRepository cl, final ClusterService cs,
+        final PrincipalExtractor principalExtractor, final PrivilegesEvaluator evaluator, ThreadPool threadPool, AuditLog auditLog) {
+        super(settings, configPath, controller, client, adminDNs, cl, cs, principalExtractor, evaluator, threadPool, auditLog);
+        this.staticNodesDnFromEsYml = settings.getAsList(ConfigConstants.OPENDISTRO_SECURITY_NODES_DN, Collections.emptyList());
+    }
+
+    @Override
+    protected void registerHandlers(RestController controller, Settings settings) {
+        if (settings.getAsBoolean(ConfigConstants.OPENDISTRO_SECURITY_NODES_DN_DYNAMIC_CONFIG_ENABLED, false)) {
+            controller.registerHandler(Method.GET, "/_opendistro/_security/api/nodesdn/{name}", this);
+            controller.registerHandler(Method.GET, "/_opendistro/_security/api/nodesdn/", this);
+            controller.registerHandler(Method.DELETE, "/_opendistro/_security/api/nodesdn/{name}", this);
+            controller.registerHandler(Method.PUT, "/_opendistro/_security/api/nodesdn/{name}", this);
```

Comment:
for PUT, PATCH and DELETE are we allowing users to modify nodesdn for the existing cluster? is there a way in the current design to distinguish between "cross_cluster" node dn and ones for the existing cluster?

Reply:
Yes, the GET has a show_all which shows the existing cluster nodesdn from YML with special key (STATIC_ES_YML_NODES_DN)

Reply:
Please reopen conv if this doesn't answer your question.

Reply:
thanks for the explanation